Custom connectors and additional typings

imatlopez opened the pull request:
Made the SentinelIterator and actual Iterator ~and made it not affect `this.options.sentinels` in place~ (tests were there to test the side effect).

Added `floatingSentinels` to the `SentinelConnectorOptions` which is called whenever `connect` is called, it must return an array of sentinels using a callback.


Welcome any feedback :)

resolves #903 for Sentinel at least :)

luin commented:
Thanks for the pull request!

It adds some overheads to add an option for this specified issue. A better way seems to be customizing SentinelConnector by overriding `SentinelConnector#connect()` to fetch dynamic endpoints first.

Currently, it could be a little bit hacky since there's no way for users to provide customized connectors to `Redis`. But it trivial to add an option to `Redis` like `connector` which will solve a broader set of problems.

imatlopez commented:
I added the `connector` property to the config, which is now where `Redis` uses its connector from. Hence if no property is set, the constructor will assign it based on the old behavior.

I reverted some of the changes to `SentinelConnector` but not all because I felt it is cleaner this way (?)

Added `AsyncSentinelConnector` as  an example of a custom connector that can be passed in.

imatlopez commented:
@luin let me know your thoughts :)

luin commented on lib/connectors/SentinelConnector/index.ts:
```diff
@@ -37,15 +41,19 @@ export interface ISentinelConnectionOptions extends ITcpConnectionOptions {
 
+export interface ISentinelConnectorOptions extends ISentinelConnectionOptions {
+  sentinels: Partial<ISentinelAddress>[]
+}
```
It seems `ISentinelConnectionOptions` isn't used other places inside the library except for `AsyncSentinelConnector`, which I think can be moved outside (like to the `examples` folder).

If people want to reuse `ISentinelConnectorOptions` with the `sentinels` property to be omitted, they can simply use `Pick<ISentinelConnectorOptions, Exclude<keyof TypeA, 'sentinels'>>`, so given that, I think we just need to keep one of the two interfaces.

luin commented on lib/connectors/AsyncSentinelConnector.ts:
```diff
@@ -0,0 +1,35 @@
+        this.fetch = fetch;
+    }
+
+    public connect(callback: NodeCallback<NetStream>, eventEmitter: ErrorEmitter) {
```
I prefer to making all public APIs use Promise instead of the outdated callback-style. So if we are going to export the connectors, the `connect` method needs to be changed to `public connect(eventEmitter: ErrorEmitter): Promise<NetStream>`.

luin commented on lib/connectors/AsyncSentinelConnector.ts:
```diff
@@ -0,0 +1,35 @@
+    private fetch: AsyncSentinelFetch
+
+    constructor(options: ISentinelConnectionOptions, fetch: AsyncSentinelFetch) {
+        options.sentinels = options.sentinels || [{host: 'localhost', port: 6379}] // Placeholder
```
What about not calling `super()` here (so that we don't need a placeholder)?

imatlopez replied:
you always have to call super though

luin commented:
Awesome work! As you said, that changes made on `SentinelConnector` make the code cleaner. The most parts of codes are great, and I've left some comments about the interface of `SentinelConnector` since this the first time we export it.

luin commented on lib/connectors/AsyncSentinelConnector.ts:
```diff
@@ -0,0 +1,35 @@
+        });
+    }
+
+}
```
I suggest moving this file to the `examples` folder since most users don't need it, and it trivial to implement one. What do you think?

imatlopez replied:
👍 now I just need to find why the tests fail

luin commented on .travis.yml:
```diff
@@ -12,6 +12,7 @@ services:
 - redis-server
 
 script:
+- npm run build
```
Why `npm run build` is needed here?

imatlopez replied:
test may have type errors like trying to access private properties, but the code itself is fine, so this guards against build time errors while the tests are for logical errors only.

imatlopez commented:
I think I dug myself a deep hole trying to change the interface to promises. I haven't figured out how to run the tests locally either so I'm gonna need some assistance 😁

imatlopez commented:
@luin I've noticed that the ESLint doesn't quite run in some parts of the code, fixing it by adding `typescript-eslint` added lots of conflicts. Once this goes in, would you like another PR adding it for code consistency? I personally like `Prettier` for code styling too.

luin commented:
@imatlopez That would be great! And Prettier is also my favorite :-)
